[Clang] Fix crash with implicit int-to-pointer conversion

ostannard opened the pull request:
If an integer is passed to the pointer argument of the __atomic_test_and_set or __atomic_clear builtins with the int-conversion error disabled or downgraded, we crashed in codegen due to assuming that the type is always a pointer after skip[ping past implicit casts.

Fixes #111293.

ostannard committed:
[Clang] Fix crash with implicit int-to-pointer conversion

If an integer is passed to the pointer argument of the
__atomic_test_and_set or __atomic_clear builtins with the int-conversion
error disabled or downgraded, we crashed in codegen due to assuming that
the type is always a pointer after skip[ping past implicit casts.

Fixes #111293.

efriedma-quic commented:
Is there some discussion somewhere of why the change from hasErrorOccurred to hasUnrecoverableErrorOccurred happened?

efriedma-quic commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -4928,8 +4928,9 @@ RValue CodeGenFunction::EmitBuiltinExpr(const GlobalDecl GD, unsigned BuiltinID,
     // Look at the argument type to determine whether this is a volatile
     // operation. The parameter type is always volatile.
     QualType PtrTy = E->getArg(0)->IgnoreImpCasts()->getType();
```
IgnoreImpCasts is usually wrong; in general, it's going to find an unrelated type.

If __atomic_test_and_set is overloaded to have both a volatile and non-volatile variant, it should be using custom type-checking in Sema.

efriedma-quic replied:
Related testcase: `void f() { volatile int x[10]; __atomic_test_and_set(x,0); }`.

shafik commented:
> Is there some discussion somewhere of why the change from hasErrorOccurred to hasUnrecoverableErrorOccurred happened?

I had asked about this in the original PR here: https://github.com/llvm/llvm-project/pull/84146 but did not get a response.

ostannard commented:
> If __atomic_test_and_set is overloaded to have both a volatile and non-volatile variant, it should be using custom type-checking in Sema.

Looking at the code again, there's a separate system used to codegen most of the atomic builtins, so I've switched to using that. That completely replaces this change, so I've started a new PR #120449, and will close this one.

> I had asked about this in the original PR here: https://github.com/llvm/llvm-project/pull/84146 but did not get a response.

I've added a comment there.
